crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -338,19 +339,32 @@ impl RecursorDnsHandle {
             }
         }
 
-        // collect missing IP addresses, select over them all, get the addresses
-        // make it configurable to query for all records?
+        // If we have no glue, collect missing IP addresses for non-child NS servers
+        // Querying for child NS servers can result in infinite recursion if the
+        // parent nameserver never returns glue records for child NS records, or does
+        // so based on cache freshness (observed with BIND)
+        //
+        // TODO: make it configurable to query for all records?
         if config_group.is_empty() && !need_ips_for_names.is_empty() {
-            debug!("need glue for {}", zone);
-            let a_resolves = need_ips_for_names.iter().take(1).map(|name| {
-                let a_query = Query::query(name.0.clone(), RecordType::A);
-                self.resolve(a_query, request_time, false).boxed()
-            });
+            debug!("need glue for {zone}");
 
-            let aaaa_resolves = need_ips_for_names.iter().take(1).map(|name| {
-                let aaaa_query = Query::query(name.0.clone(), RecordType::AAAA);
-                self.resolve(aaaa_query, request_time, false).boxed()
-            });
+            let a_resolves = need_ips_for_names
+                .iter()
+                .filter(|name| !crate::is_subzone(zone.clone(), name.0.clone()))
```

Comment:
This was probably missed in a previous review, I think the signature of `is_subzone` is incorrect, and it would be good for us to fix this so these clones aren't necessary.

Could you as part of this change change `is_subzone` from `fn is_subzone(parent: Name, child: Name)` to `fn is_subzone(parent: &Name, child: &Name)`? I don't see any reason why that shouldn't be valid, and would remove these clones.

Reply:
looks like @djc fixed this in #2334

Reply:
I believe @djc put in a separate PR for this a few days ago. I can move that change into this PR if you prefer.